### PR TITLE
feat(kuma-cp): don't create default policies by default and use targetRef

### PR DESCRIPTION
**Pull request by lukidzi**
### Checklist prior to review

Why is KUMA_DEFAULTS_CREATE_MESH_ROUTING_RESOURCES necessary? 
- In situations where `Global` is newer than the zone, the zone might not be able to route traffic due to the absence of default policies. To facilitate an easier migration process, users can modify the default behavior exclusively within the global or zone (federated) settings.

Most of the changes are test files. There is an issue with gateway timeouts. Previously, the gateway code was setting specific timeouts for the gateway but since we created the default MeshTimeout policy it overrides the defaults. So far we cannot create a default policy just for gateways so we might need to document how the policy should look like and what good timeouts are for the gateway.

- [X] [Link to relevant issue][1] as well as docs and UI issues -- fix: https://github.com/kumahq/kuma/issues/6592
- [X] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [X] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [X] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [X] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin.go [hunk 1]
```diff
@@ -119,6 +123,9 @@ func ApplyToGateway(
 	rules []ToRouteRule,
 ) error {
 	var limits []plugin_gateway.RuntimeResoureLimitListener
+	if len(rules) == 0 {
+		return nil
+	}
 
 	for _, info := range plugin_gateway.ExtractGatewayListeners(proxy) {
```
Comment (lukidzi) [inline pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin.go:130]: This function `ExtractGatewayListeners` reuses some old part of gateway code which has services filtered by the traffic permissions

File: test/e2e_env/universal/grpc/grpc.go [hunk 2]
```diff
@@ -62,7 +62,7 @@ func GRPC() {
 		}, "30s", "1s").Should(Succeed())
 	})
 
-	It("MeshHTTPRoute DOES NOT split the traffic between two gRPC services", func() {
+	It("MeshHTTPRoute does split the traffic between two gRPC services", func() {
```
Comment (slonka) [inline test/e2e_env/universal/grpc/grpc.go:65]: why is this now the case? why is default policy creation tied to MeshHTTPRoute splitting gRPC traffic?

File: test/e2e_env/universal/grpc/grpc.go [hunk 3]
```diff
@@ -96,9 +96,8 @@ spec:
 				universal.Cluster, "second-test-server", "http://localhost:9901/stats?format=prometheus",
 			)
 			g.Expect(err).ToNot(HaveOccurred())
-			// todo switch .ToNot to .To when https://github.com/kumahq/kuma/issues/3325 is implemented
```
Comment (slonka) [inline test/e2e_env/universal/grpc/grpc.go:99]: same question as above

**Review by slonka — COMMENTED** [review]
Added some comments and have a question about this:

> Why is KUMA_DEFAULTS_CREATE_MESH_RESOURCES necessary?
>
> In situations where Global is never set as a zone, the zone might not be able to route traffic due to the absence of default policies. To facilitate an easier migration process, users can modify the default behavior exclusively within the global or zone (federated) settings.

`In situations where Global is never set as a zone` -> what do you mean by this? you mean a non federated zone?

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go [hunk 4]
```diff
@@ -90,6 +89,12 @@ func generateListeners(
 		filterChainBuilder.
 			Configure(envoy_listeners.AddFilterChainConfigurer(outboundRouteConfigurer))
 
+		// TODO: https://github.com/kumahq/kuma/issues/3325
```
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go:92]: Agree, this should really be a different PR

[the remark this answers is not in the document]
Comment (lukidzi) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go:92]: I agree but currently, we don't have the policy `MeshGRPCRoute` and without this change we won't be able to fully switch to new policies and still partially relies on TrafficRoute for grpc traffic.

File: pkg/defaults/mesh/mesh.go [hunk 5]
```diff
@@ -52,44 +53,45 @@ func EnsureDefaultMeshResources(
...
-	}
-
-	defaultResourceBuilders := map[string]func() model.Resource{
-		"allow-all":           defaultTrafficPermissionResource,
-		"route-all":           defaultTrafficRouteResource,
-		"timeout-all":         DefaultTimeoutResource,
-		"circuit-breaker-all": defaultCircuitBreakerResource,
-		"retry-all":           defaultRetryResource,
-	}
+	if createMeshDefaultResources {
```
[the remark this answers is not in the document]
Comment (lukidzi) [reply pkg/defaults/mesh/mesh.go:56]: I think that is correct because that is a new behavior that doesn't require default policies.

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go [hunk 6]
```diff
@@ -167,19 +171,14 @@ func prepareRoutes(
...
-				}
-			}
 			route.BackendRefs = []common_api.BackendRef{
 				{
-					TargetRef: targetRef,
-					Weight:    pointer.To(uint(100)),
+					TargetRef: common_api.TargetRef{
+						Kind: common_api.MeshService,
+						Name: serviceName,
+						Tags: tags,
```
[the remark this answers is not in the document]
Comment (lukidzi) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go:179]: Previously, we only required the `mesh` tag. However, to support virtual outbound, we now require additional tags.

File: test/e2e_env/universal/grpc/grpc.go (hunk 2 above)
Comment (lukidzi) [reply test/e2e_env/universal/grpc/grpc.go:65]: To fully switch to targetRef policies we need grpc support and because of the lack of MeshGRPCRoute we can currently support them with MeshHTTPRoute

File: pkg/defaults/mesh/mesh_test.go [hunk 7]
```diff
@@ -30,9 +30,43 @@ var _ = Describe("EnsureDefaultMeshResources", func() {
...
+		// when
+		err := mesh.EnsureDefaultMeshResources(context.Background(), resManager, defaultMesh, []string{}, context.Background(), false)
+		Expect(err).ToNot(HaveOccurred())
+
+		// and Dataplane Token Signing Key for the mesh exists
+		err = resManager.Get(context.Background(), system.NewSecretResource(), core_store.GetBy(tokens.SigningKeyResourceKey(issuer.DataplaneTokenSigningKeyPrefix(model.DefaultMesh), tokens.DefaultKeyID, model.DefaultMesh)))
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	It("should not create default policies", func() {
```
Comment (michaelbeaumont) [inline pkg/defaults/mesh/mesh_test.go:43]: kind of a nit but IMO the test case could be named better, like reading this `It` and the next `It` are just confusing together since they're opposite. I guess this one should be "should not create... if resources are already present"?

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/gateway.go [hunk 8]
```diff
@@ -37,7 +37,12 @@ func generateGatewayListeners(
 	if info.Listener.CrossMesh {
 		protocol = mesh_proto.MeshGateway_Listener_HTTPS
 	}
-	res, filterChainBuilders, err := FilterChainGenerators[protocol].Generate(ctx, info, gatewayHosts)
+	filterGen, found := FilterChainGenerators[protocol]
+	if !found {
```
Comment (michaelbeaumont) [inline pkg/plugins/policies/meshhttproute/plugin/v1alpha1/gateway.go:41]: Is this the gRPC thing or why is this necessary?

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/gateway_routes.go [hunk 9]
```diff
@@ -17,7 +17,9 @@ import (
 
 func GenerateEnvoyRouteEntries(host plugin_gateway.GatewayHost, toRules []ToRouteRule) []route.Entry {
 	var entries []route.Entry
-
+	if len(toRules) == 0 {
```
Comment (michaelbeaumont) [inline pkg/plugins/policies/meshhttproute/plugin/v1alpha1/gateway_routes.go:20]: I tihnk this can be removed? The check already happens in `plugin.go` now

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go:179]: Can we please make different PRs for this in the future, it will now be significantly harder to figure out why this change was made after the fact.

File: pkg/plugins/runtime/gateway/gateway_route_generator_test.go [hunk 10]
```diff
@@ -1606,13 +1597,7 @@ conf:
       - destination:
           kuma.io/service: echo-service
 `,
-			WithoutResource{
-				Resource: core_mesh.TimeoutType,
-				Mesh:     "default",
-				Name:     "timeout-all-default",
```
Comment (michaelbeaumont) [inline pkg/plugins/runtime/gateway/gateway_route_generator_test.go:1612]: How exactly should these tests be changing, don't we kind of need both cases? So for example here we now need to _add_ `Timeout` resources?

File: pkg/plugins/runtime/gateway/testdata/http/no-timeout.yaml [hunk 11]
```diff
@@ -25,7 +25,7 @@ Clusters:
         envoy.extensions.upstreams.http.v3.HttpProtocolOptions:
           '@type': type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions
           commonHttpProtocolOptions:
-            idleTimeout: 0s
+            idleTimeout: 3600s
```
Comment (michaelbeaumont) [inline pkg/plugins/runtime/gateway/testdata/http/no-timeout.yaml:28]: Why is this changing?

**Comment by lukidzi** [comment]
> Added some comments and have a question about this:
> 
> > Why is KUMA_DEFAULTS_CREATE_MESH_RESOURCES necessary?
> > In situations where Global is never set as a zone, the zone might not be able to route traffic due to the absence of default policies. To facilitate an easier migration process, users can modify the default behavior exclusively within the global or zone (federated) settings.
> 
> `In situations where Global is never set as a zone` -> what do you mean by this? you mean a non federated zone?

I fixed the comment i meant newer than zone

File: pkg/plugins/policies/meshtcproute/plugin/v1alpha1/listeners.go [hunk 12]
```diff
@@ -27,6 +27,9 @@ func generateListeners(
 	// For one outbound we pick one traffic route, so LB and Timeout are
 	// the same.
 	clusterCache := map[common_api.TargetRefHash]string{}
+	if proxy == nil || proxy.Dataplane == nil || proxy.Dataplane.Spec.GetNetworking() == nil {
```
Comment (michaelbeaumont) [inline pkg/plugins/policies/meshtcproute/plugin/v1alpha1/listeners.go:30]: when can `networking` be `nil`?

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/gateway.go (hunk 8 above)
Comment (lukidzi) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/gateway.go:41]: Yes, I had a nil pointer in case of unknown protocol

File: pkg/plugins/policies/meshtcproute/plugin/v1alpha1/listeners.go (hunk 12 above)
Comment (lukidzi) [reply pkg/plugins/policies/meshtcproute/plugin/v1alpha1/listeners.go:30]: Only in case of Egress/Ingress

File: pkg/plugins/runtime/gateway/gateway_route_generator_test.go (hunk 10 above)
Comment (lukidzi) [reply pkg/plugins/runtime/gateway/gateway_route_generator_test.go:1612]: Reverted change, I am going to switch in the next PR and add default policy

File: pkg/plugins/runtime/gateway/testdata/http/no-timeout.yaml (hunk 11 above)
Comment (lukidzi) [reply pkg/plugins/runtime/gateway/testdata/http/no-timeout.yaml]: This value changes because the gateway code is setting a different timeout. Because MeshTimeout is a new default it overrides the timeouts set by gateway code. There is no way (yet) to create only a policy that targets all gateways

File: pkg/config/app/kuma-cp/config.go [hunk 13]
```diff
@@ -380,6 +380,13 @@ func DefaultGeneralConfig() *GeneralConfig {
 	}
 }
 
+func DefaultDefaultsConfig() *Defaults {
+	return &Defaults{
+		SkipMeshCreation:    false,
+		CreateMeshResources: true,
```
[the remark this answers is not in the document]
Comment (lukidzi) [reply pkg/config/app/kuma-cp/config.go]: In the 2nd PR because might change some more tests

**Comment by lukidzi** [comment]
[closing in favor of ](https://github.com/kumahq/kuma/pull/8920)
